src/main/java/io/gravitee/policy/generatehttpsignature/GenerateHttpSignaturePolicy.java

```diff
@@ -0,0 +1,161 @@
+import java.io.IOException;
+import java.security.Key;
+import java.util.ArrayList;
+import java.util.List;
+import javax.crypto.spec.SecretKeySpec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.tomitribe.auth.signatures.Signature;
+import org.tomitribe.auth.signatures.Signer;
+
+/**
+ * @author Yann TAVERNIER (yann.tavernier at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class GenerateHttpSignaturePolicy {
+
+    private static final Logger logger = LoggerFactory.getLogger(GenerateHttpSignaturePolicy.class);
+
+    private static final String HTTP_SIGNATURE_IMPOSSIBLE_GENERATION = "HTTP_SIGNATURE_IMPOSSIBLE_GENERATION";
+    private static final String ERROR_MESSAGE = "Unable to generate HTTP Signature:";
+
+    private final GenerateHttpSignaturePolicyConfiguration configuration;
+
+    public GenerateHttpSignaturePolicy(GenerateHttpSignaturePolicyConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    @OnRequest
+    public void onRequest(Request request, Response response, ExecutionContext context, PolicyChain chain) {
+        HttpHeaders requestHeaders = new HttpHeaders(request.headers());
+        List<String> configuredHeaders = new ArrayList<>(configuration.getHeaders());
+
+        final String checkHeadersErrorMessage = checkHeaders(requestHeaders, configuredHeaders, chain);
+        if (checkHeadersErrorMessage != null) {
+            logger.warn(checkHeadersErrorMessage);
+            chain.failWith(PolicyResult.failure(HTTP_SIGNATURE_IMPOSSIBLE_GENERATION, 400, checkHeadersErrorMessage));
+            return;
+        }
+
+        final Signature signatureFromConfiguration = buildSignatureFromConfiguration(context, request, configuredHeaders);
+
+        final Signer signer = buildSigner(context, signatureFromConfiguration);
+        final Signature signature;
+
+        try {
+            signature = signer.sign(request.method().name().toLowerCase(), request.path(), requestHeaders.toSingleValueMap());
+        } catch (IOException e) {
+            final String errorMessage = String.format("%s %s", ERROR_MESSAGE, e.getMessage());
+            logger.warn(errorMessage);
+            request.metrics().setMessage(e.getMessage());
+            chain.failWith(PolicyResult.failure(HTTP_SIGNATURE_IMPOSSIBLE_GENERATION, 400, errorMessage));
+            return;
+        }
+
+        setSignatureHeader(request, signature);
+
+        chain.doNext(request, response);
+    }
+
+    /**
+     * Validate the request by verifying it has the headers mentioned in policy's configuration.
+     * @param headers the incoming request's headers.
+     * @param configuredHeaders the headers from policy's configuration.
+     *
+     */
+    String checkHeaders(HttpHeaders headers, List<String> configuredHeaders, PolicyChain chain) {
+        if (!isEmpty(configuredHeaders)) {
+            if (!headers.keySet().containsAll(configuredHeaders)) {
```

Comment:
Should we do a case-insensitive headers comparisons in this function ?
configuredHeaders from configuration can contains different case as headers from incoming request ?

Reply:
good point. I think it's indeed a good idea to do it systematically

Reply:
HttpHeaders are, by nature, case-insensitive. So I would prefer to create a new method in HttpHeaders to check that it `containsAll` specified header keys.

Reply:
done!